[FSDP] Fix collective mismatch error with conditional parameter usage.

mansiag05 opened the pull request:
When using FSDP with conditional parameter usage (e.g., different ranks using different parameters due to model architecture or control flow), the post_backward gradient reduction would fail with:

  RuntimeError: Detected mismatch between collectives on ranks.

This occurred because different ranks had different sets of parameters with gradients, causing mismatched reduce-scatter operations.

Root Cause:
The FSDPParamGroup.post_backward() method only collected parameters that had gradients. When ranks had divergent parameter usage, they participated in collectives with different tensor counts/shapes.

Solution:
Modified post_backward() to include all parameters in gradient reduction by adding zero-filled gradients for parameters without gradients. This ensures all ranks have consistent tensor shapes during collective operations while maintaining correct gradient values.

Fixes #158719

mansiag05 committed:
[FSDP] Fix collective mismatch error with conditional parameter usage.

When using FSDP with conditional parameter usage (e.g., different ranks
using different parameters due to model architecture or control flow),
the post_backward gradient reduction would fail with:

  RuntimeError: Detected mismatch between collectives on ranks.

This occurred because different ranks had different sets of parameters
with gradients, causing mismatched reduce-scatter operations.

Root Cause:
The FSDPParamGroup.post_backward() method only collected parameters
that had gradients. When ranks had divergent parameter usage, they
participated in collectives with different tensor counts/shapes.

Solution:
Modified post_backward() to include all parameters in gradient
reduction by adding zero-filled gradients for parameters without
gradients. This ensures all ranks have consistent tensor shapes
during collective operations while maintaining correct gradient values.

weifengpy reviewed (changes requested):
for fine-tuning, most params are frozon except lora adapters. reduce-scattering dummy zeros could be a big perf hit. 

could I have a practical example that a parameters on rank0 requires grad but does not require grad on rank1?

fsdp2 assumes spmd. so a param should either require grad on all ranks or does not require grad on all ranks

weifengpy commented:
> Practical example: we train on observation/measurement data from collections of satellites with each batch sample being a measurement at a specific point in time from multiple satellites. The data for individual satellites is not available continuously (random satellite outages/maintenance for short periods, problems with communication). This causes rank0 and rank1 to have different gradient flows, e.g. the some embedding networks might not be used for a specific sample since the corresponding satellites did not provide data for this time instance.

I understand the point now. thanks for explaining :)

this PR introduces dummy reduce-scatters by default. we need the new behavior to be simple and explicit, without regressing existing perfs. eg `FSDPModule.set_dummy_reduce_scatter(True/False)`

ashutoshuiuc commented:
this PR is also needed for some MoE architectures/models (like `Qwen/Qwen3-30B-A3B`) where gradients are not backpropped by ranks for non-active experts which causes deadlock in optimization step.

Let me know if I can help accelerate the release of this fix.

mansiag05 committed:
Add an opt-in FSDPModule.set_reduce_scatter_unused_params() API that
  pads zero gradients for parameters without gradients, ensuring all ranks
  issue matching collectives. This is analogous to DDP's
  find_unused_parameters and is off by default to avoid perf regression.

mansiag05 committed:
Correcting the lint Error.

mansiag05 committed:
[FSDP] Fix reduce_scatter_unused_params to skip frozen params and use correct grad dtype

When reduce_scatter_unused_params is enabled, skip parameters that don't
require gradients (frozen params) to avoid unnecessary zero-grad padding.
Also let zeros_like infer dtype from the parameter itself rather than
manually computing grad_dtype.

mansiag05 committed:
Correcting lint.

weifengpy reviewed (dismissed):
approved. thanks for addressing the feedback!

stmcgovern commented:
Hi @weifengpy any updates?

weifengpy commented:
> Hi @weifengpy any updates?

@stmcgovern will get back to this by Friday

weifengpy commented:
> > Hi @weifengpy any updates?
> 
> @stmcgovern will get back to this by Friday

internal change pending on approval - should have answer soon early next week

weifengpy commented:
> internal change pending on approval - should have answer soon early next week

@stmcgovern @mansiag05 I have landed the internal change. You can rebase this PR on latest pytorch and try land again. Let me know if you need CI approval

mansiag05 commented:
Hello @weifengpy, could you please provide your approval to the PR again.

weifengpy commented:
> Hello @weifengpy, could you please provide your approval to the PR again.

approved

mansiag05 commented:
@pytorchbot merge

weifengpy commented:
@pytorchmergebot -i

weifengpy commented:
@pytorchmergebot merge -i

weifengpy commented:
@pytorchmergebot merge -f "unrelated CI error"

Irvingwangjr commented:
<img width="3394" height="1012" alt="image" src="https://github.com/user-attachments/assets/ec71fb78-cfa3-49a2-80a5-374933986c84" />

I encountered this problem when using EP+FSDP. It looks like unshard_param corresponds to the expert's EP DTensor, and then the tensor that comes out of zero is also a DTensor. However, foreach_reduce_scatter_copy_in(unsharded_grads, reduce_scatter_input, world_size) — the reduce_scatter_input will be a plain tensor buffer.

weifengpy commented:
> then the tensor that comes out of zero is also a DTensor.

@Irvingwangjr  thanks for reporting the issue! I added you as co-author for fix https://github.com/pytorch/pytorch/pull/183040 . Are you interested in signing EasyCLA so I can land it co-authored ? https://github.com/pytorch/pytorch/pull/183040#issuecomment-4411865365

Irvingwangjr commented:
> > then the tensor that comes out of zero is also a DTensor.
> 
> @Irvingwangjr thanks for reporting the issue! I added you as co-author for fix #183040 . Are you interested in signing EasyCLA so I can land it co-authored ? [#183040 (comment)](https://github.com/pytorch/pytorch/pull/183040#issuecomment-4411865365)

thx!
